python/desc/imsim/skyModel.py

```diff
@@ -155,62 +180,350 @@ def addNoiseAndBackground(self, image, bandpass=None, m5=None,
+    This is a subclass of ESOSkyModel and applies the galsim.Silicon
+    sensor model to the sky background photons derived from the ESO
+    sky brightness model.
+    """
+    def __init__(self, obs_metadata, photParams, seed=None, bandpassDict=None,
+                 logger=None):
+        """
+        Parameters
+        ----------
+        obs_metadata: lsst.sims.utils.ObservationMetaData
+            Visit-specific data such as pointing direction,
+            observation time, seeing, bandpass info, etc..  This info
+            is extracted from the phosim instance catalog headers.
+        photParams: lsst.sims.photUtils.PhotometricParameters
+            Visit-specific photometric parameters, such as exposure time, gain,
+            bandpass, etc..
+        seed: int, optional
+            Seed value passed to the random number generator used by
+            the noise model. Defaults to None, which causes GalSim to
+            generate the seed from the system.
+        bandpassDict: lsst.sims.photUtils.BandpassDict, optional
+            Bandpass dictionary used by the sims code.  If None (default),
+            the BandpassDict.loadBandpassesFromFiles function is called
+            which reads in the standard LSST bandpasses.
+        """
+        super(ESOSiliconSkyModel, self).__init__(obs_metadata, photParams,
+                                                 seed=seed,
+                                                 bandpassDict=bandpassDict,
+                                                 addNoise=addNoise,
+                                                 addBackground=addBackground,
+                                                 logger=logger)
+
+        # Wavelength and angle samplers need only be constructed at
+        # most once per SED and bandpass, so build them lazily as
+        # properties.
+        self._waves = None
+        self._angles = None
+
+    @property
+    def waves(self):
+        if self._waves is None:
+            sed = galsim.SED(galsim.LookupTable(x=self.skyModel.wave,
+                                                f=self.skyModel.spec[0,:]),
+                             wave_type='nm', flux_type='flambda')
+            bandPassName = self.obs_metadata.bandpass
+            bandpass = self.bandpassDict[bandPassName]
+            index = np.where(bandpass.sb != 0)
+            gs_bandpass \
+                = galsim.Bandpass(galsim.LookupTable(x=bandpass.wavelen[index],
+                                                     f=bandpass.sb[index]),
+                                  wave_type='nm')
+            self._waves = galsim.WavelengthSampler(sed=sed,
+                                                   bandpass=gs_bandpass,
+                                                   rng=self.randomNumbers)
+        return self._waves
+
+    @property
+    def angles(self):
+        if self._angles is None:
+            fratio = 1.234  # From https://www.lsst.org/scientists/keynumbers
```

Comment:
In the future, should we add `fratio` and `obscuration` into `sims.photUtils.PhotometricParameters`, so that ImSim can be extended to non-LSST systems?

Reply:
Yes, this is a sensible suggestion.